core/dbt/parser/base.py

```diff
@@ -306,6 +311,19 @@ def update_parsed_node_config(
             else:
                 parsed_node.docs = Docs(show=docs_show)
 
+        # If we have constraints_enabled in the config, copy to node level, for backwards
+        # compatibility with earlier node-only config.
+        if config_dict.get("constraints_enabled", False):
+            parsed_node.constraints_enabled = True
+
+            parser_name = type(self).__name__
+            if parser_name == "ModelParser":
+                original_file_path = parsed_node.original_file_path
+                error_message = "\n    `constraints_enabled=true` can only be configured within `schema.yml` files\n      NOT within a model file(ex: .sql, .py) or `dbt_project.yml`."
+                raise CompilationException(
```

Comment:
This should actually be a ParsingException. Ideally exceptions generated at parse time should all be ParsingExceptions and not CompilationExceptions, but it was tricky to set up the distinction in the Jinja code so it hasn't been completed yet.